### PR TITLE
Release 2.0

### PR DESCRIPTION
**Pull request by tkohout**
- Updated Swinject to 2.0
- Migrated to new swinject
- Resolved https://github.com/Swinject/SwinjectAutoregistration/issues/2 Added warnings when
   - fails to resolve service with optional dependencies
   - fails to resolve service with implicitly unwrapped dependencies
   - fails to resolve service with more than maximum supported number of dependencies
   - trying to register service with same type of dynamic arguments
- Added assert tests https://github.com/Swinject/SwinjectAutoregistration/issues/6
- Added swift package manager https://github.com/Swinject/SwinjectAutoregistration/issues/9

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,14 +1,39 @@
...
+      before_install:
+        - eval "$(curl -sL https://gist.githubusercontent.com/kylef/5c0475ff02b7c7671d2a/raw/9f442512a46d7a2af7b850d65a7e9bd31edfb09b/swiftenv-install.sh)"
+      script:
+        - mv .Package.test.swift Package.swift
+        - swift build
+        - swift test
+    - env: JOB="POD_LINT"
+      script:
+        - pod lib lint
+    - env: JOB="XCODE" DEST="OS=8.4,name=iPhone 5" SCHEME="SwinjectAutoregistration-iOS" SDK="iphonesimulator" ACTION="test"
```
Comment (yoichitgy) [inline .travis.yml]: Changing the device to iPhone 5s looks ok to fix the error on Travis with Nimble.
https://travis-ci.org/Swinject/SwinjectAutoregistration/jobs/196326139

```
DEST="OS=8.4,name=iPhone 5s" 
```

**Comment by yoichitgy** [comment]
It looks good to run Carthage command with `--use-submodules` option to add dependencies as git submodules not to commit files of Swinject, Quick and Nimble to this repository. It's easier to review a pull request without those library files.

@tkohout, using the option is ok for you?

**Comment by jakubvano** [comment]
👍 Good job guys on 2.0 release!
@yoichitgy Is there a benefit of having carthage dependencies included as submodules and subprojects as opposed to just linked frameworks (not pushed in repo)? I suspect that when you have both `Swinject` and `SwinjectAutoregistration` in project Cartfile, this causes carthage effectively build `Swinject` twice (once as a carthage dependency, once as a subproject in `SwinjectAutoregistration`), and also might be causing [Swinject#184](https://github.com/Swinject/Swinject/issues/184).

**Comment by yoichitgy** [comment]
Thanks @jakubvano for pointing out the issue👍👍👍 It looks good not to commit submodules.

Or maybe users should add only SwinjectAutoregistration to Cartfile to install both SwinjectAutoregistration and  Swinject??? For example, installation guide of [ReactiveCocoa](https://github.com/ReactiveCocoa/ReactiveCocoa) tells to add only ReactiveCocoa to Cartfile to install with ReactiveSwift. Are you familiar with this idea?

I'm going to try some experiments.

**Comment by jakubvano** [comment]
I get what you mean, but this approach would not cover the case when you use more than one Swinject extension (e.g. SwinjectAutoregistration and SwinjectStoryboard).

**Comment by yoichitgy** [comment]
I agree. My approach doesn't fit in the case using more than a Swinject extension. I'm going to have some experiment [as @jakubvano said](https://github.com/Swinject/SwinjectAutoregistration/pull/11#issuecomment-276601787).

**Comment by yoichitgy** [comment]
I started adding depending projects to a project because of this issue:
https://github.com/Swinject/Swinject/issues/40

For example, Quick and Nimble were added as projects, not prebuilt frameworks. In the same way, I added Swinject project to SwinjectStoryboard and SwinjectPropertyLoader.

The original issue looked already resolved, so we have no reasons to add depending projects. As @jakubvano told, we can just add frameworks built by Carthage, and depending projects can be removed.

**Comment by yoichitgy** [comment]
I made the migration to prebuilt frameworks in the following branch. I think it should be in another pull request, so that I'm going to just review this current pull request.
https://github.com/Swinject/SwinjectAutoregistration/tree/migrate-to-prebuilt-frameworks

File: Sources/CheckResolved.swift [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+func unresolvedService<A>(_ a: A?) -> String? {
+   return ( a == nil ? "\(A.self)" : nil )
+}
+
+func checkResolved<Service, A>(initializer: (A) -> Service, services a: A?){
+   let unresolved = ( [a] as [Any?] ).filter { $0 == nil }
+   if unresolved.count > 0 {
+       let warningsMessage = warnings(forInitializer: initializer).map { "\($0.message)\n" }.joined()
+       fatalError("SwinjectAutoregistration: Resolution failed.\n\(warningsMessage)Unresolved service: \(unresolvedService(a)!)\nInitializer: (\(A.self)) -> \(Service.self)")
```
Comment (yoichitgy) [inline Sources/CheckResolved.swift]: If this is just a warning message, shouldn't this be `print` or `NSLog`? The other warnings in this file are the same.

Comment (tkohout) [inline Sources/CheckResolved.swift]: Well the warnings describe what caused that the service was unresolved: Something like:
```
⚠ Autoregistration of implicitly unwrapped optional dependency (DependencyA)! is not supported. Use regular `register` method. 
```
And if i would just print them before the fatalError they might have been overlooked

Comment (tkohout) [inline Sources/CheckResolved.swift]: The tvOS throw assertion fail is fixed

Comment (yoichitgy) [inline Sources/CheckResolved.swift]: Thank you for fixing the tvOS issue!

[nits] This is just a tiny thing, but if `warningMessage` is used for `fatalError` naming the parameter `errorMessage` sounds less misunderstandable. Maybe related parameter names also sound good with "error"😉

Comment (tkohout) [inline Sources/CheckResolved.swift]: Thanks, I refactored the code to address your notes.

**Comment by yoichitgy** [comment]
I think we can release v2.0.0 now.

My changes to migrate to prebuilt frameworks can be released as v2.1 or v2.0.1 because the changes are not considered as breaking changes.
https://github.com/Swinject/SwinjectAutoregistration/tree/migrate-to-prebuilt-frameworks

**Pull request merged**